### ISSUE TITLE
Non-rooted expressions support

### ISSUE THREAD

**Issue by davidkhess** [jmespath/jmespath.py#110]
I may just be missing it, but it doesn't seem to be possible to do the equivalent of `//` in XPath. I.e. find the desired expression anywhere in the supplied JSON versus from the root of it. If not, can that be made a feature request?

The use case is when you are trying to find subsets of JSON in a document whose structure you cannot predict in advance.

**Comment by jamesls** [comment]
Not currently possible, but something I think would be useful.

**Comment by davidkhess** [comment]
It's pretty necessary if you are working with JSON with an unknown schema but need to recognize subsets within it. I find it crops up when using JSON for metadata.

I'm not sure how many folks need this – but it's a deal breaker for that kind of situation.

**Comment by gholder** [comment]
Agreeing with DavidHess.  
Your documentation and the project look great, but it's necessary to do the non-rooted searches.

**Comment by EricLanduyt** [comment]
I agree too, non-rooted expressions is the only real thing missing in JMESPath.

**Comment by caphrim007** [comment]
@jamesls would you propose a syntax for this to the JMESPath specification? even if it's not yet implemented, a proposed syntax would allow someone to produce a PR that does implement it for your review

**Comment by sauron918** [comment]
It's really missing.. Is there any changes in this direction since 2016?

**Comment by ogaida** [comment]
for the meantime, here is a bash function i use to search values in json with jq:

```
function jsonValueSearch (){
    searchString=$1;
    jq -r 'paths as $path | select(getpath($path) == "'"$searchString"'") | $path | ".\(map("[\(tojson)]") | join(""))"'
}
```

**Comment by chipsenkbeil** [comment]
@jamesls what are your thoughts on adopting the same recursive descent syntax as JSONPath and [E4X](https://en.wikipedia.org/wiki/ECMAScript_for_XML), where `..` is used to signify the descent operator.

I was going to switch from a Rust JSONPath library to the jmespath.rs library (after getting it modernized), but I need recursive descent for my use case. I serialize a lot of Rust data structures representing the vimwiki markup language into JSON. If I wanted to look up a tag, snippet of text, etc. I only know the subtree within JSON and not necessarily the full path.

**Comment by springcomp** [comment]
@chipsenkbeil We want to [propose this feature to JMESPath](https://github.com/jmespath/jmespath.jep/issues/31) and will include this in the [next iteration for JMESPath Community](https://github.com/jmespath-community/jmespath.spec/discussions/151).

**Comment by dharasty** [comment]
- 9:30am: went searching for Python module to let me "query" dictionaries/JSON objects
- 9:45am: happy to fine JMESPath!
- 9:46am: pip install jmespath
- 10:10am: highly surprised that "non-rooted" searches are not supported
- 10:11am: determined this is not fit for my purpose
- 10:12am: pip uninstall jmespath

-- me, today
